Allow DataGrid's columns popover to optionally not allow hiding and/or reordering

chandlerprall opened the pull request:
### Summary

Fixes #2622 

![example popover](https://d.pr/i/bwE4eq.png)

* changed text of the column popover button from "Hide fields" / "fields hidden" to "Columns" / "columns hidden". Completely open to further changes. Also, the icon doesn't make sense now @snide 
* extended `toolbarVisibility.showColumnSelector` to allow objects as well. I think this nicely and naturally extends the existing design of the `toolbarVisibility` value, where false values disable, undefined/true values enable, and an object customizes further
 * created `allowHide` and `allowReorder` values on `toolbarVisibility.showColumnSelector`; this breaks with the existing `show*` pattern for naming. Do we care?

@snide would you mind doing a design pass?

### Checklist

~- [ ] Check against **all themes** for compatibility in both light and dark modes~
~- [ ] Checked in **mobile**~
~- [ ] Checked in **IE11** and **Firefox**~
- [x] Props have proper **autodocs**
- [x] Added **documentation** examples
- [x] Added or updated **jest tests**
- [x] Checked for **breaking changes** and labeled appropriately
- [x] Checked for **accessibility** including keyboard-only and screenreader modes
- [x] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2993/

myasonik commented:
Maybe we need a few states for this? What are y'all's thoughts on:

1. Only sort
	- Icon: a horizontal version of the sort icon
	- Text: "Sort columns" / "𝑛 columns sorted"

2. Only visibility controls
	- Icon: closed eye (what we have today)
	- Alternate icon proposal: could also abuse the kqlOperand or merge icons to serve this purpose
	- Text: "Hide fields" / "𝑛 fields hidden" (what we have today)
	- Alternate text proposal: use "columns" instead of "fields" to keep this button always saying "columns"

3. Both options available
	- Icon: controls (horizontal or vertical)
	- Text: "Configure columns" (no "configured" text)
	- Tooltip option: Could show a tooltip on the button with "𝑛 columns sorted" and/or "𝑛 fields hidden" as needed

myasonik commented on src/components/datagrid/column_selector.tsx:
```diff
@@ -86,7 +108,7 @@ export const useColumnSelector = (
             'euiColumnSelector.buttonActive',
           ]}
-          defaults={['Hide fields', 'fields hidden']}>
+          defaults={['Columns', 'columns hidden']}>
```
Do we need to improve the pluralization here? Just realized it's a little odd when it says "1 columns hidden"...

myasonik commented:
> created allowHide and allowReorder values on toolbarVisibility.showColumnSelector; this breaks with the existing show* pattern for naming. Do we care?

I think it's fine. If we decide we do care, I'd propose actually changing all the existing `show*` props to be `allow*` props. Just glancing through the docs, it seems like it would work for all of them.

thompsongl commented:
I'm going to wait for @snide to do his design pass before reviewing. Especially given @myasonik's ideas, which could change the code a bit if implemented.

snide commented:
Pushed some design changes. Fixed the various states and icons.

Whether or not you can or can not hide columns, we want to keep this button fairly singular in focus. So I kept the "columns" naming and changed the icon to something that works a little more generically. I also fixed the pluralization. Doing a quick docs change, but this should be safe for review now.

I also fixed the styling changes when the label is not used.

snide commented:
Screens for the lazy

![image](https://user-images.githubusercontent.com/324519/76355339-5a3f3480-62d1-11ea-82a9-00a7e2cc09ea.png)

![image](https://user-images.githubusercontent.com/324519/76355351-60351580-62d1-11ea-8d0f-17f427229edf.png)

![image](https://user-images.githubusercontent.com/324519/76355359-64613300-62d1-11ea-9c90-955b2f72a4d2.png)

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2993/

myasonik commented on src/components/datagrid/column_selector.tsx:
```diff
@@ -81,48 +122,39 @@ export const useColumnSelector = (
+          className={controlBtnClasses}
+          data-test-subj="dataGridColumnSelectorButton"
+          onClick={() => setIsOpen(!isOpen)}>
+          {numberOfHiddenFields > 0 ? numberOfHiddenFields : null} {buttonText}
```
I'm not sure how EUI's i18n system works (or, how it integrates with Kibana) but should we pass the value into the translation string?

Different languages might put the number in a different position or might add a certain count word somewhere in the string as well so we usually want to keep the number and the phrase together.

snide commented:
The docs are now updated. I moved the demo for this into the toolbar section, which is where I think it belongs. I have a nested state toggle in the popover.

![](http://snid.es/276a32cff6ae/Screen%252520Recording%2525202020-03-10%252520at%25252003.11%252520PM.gif)

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2993/

chandlerprall commented:
Added another nested toggle to the docs for allowing/disabling column reordering, next to the one Dave added for column show/hide capability.

Updated the I18n stuff to place the # of columns hidden value programatically.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2993/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2993/
